x-pack/plugins/alerting/server/task_runner/task_runner.ts

```diff
@@ -589,6 +596,62 @@ export class TaskRunner<
   }
 }
 
+interface TrackAlertDurationsParams<
+  InstanceState extends AlertInstanceState,
+  InstanceContext extends AlertInstanceContext
+> {
+  originalAlerts: Dictionary<AlertInstance<InstanceState, InstanceContext>>;
+  currentAlerts: Dictionary<AlertInstance<InstanceState, InstanceContext>>;
+  recoveredAlerts: Dictionary<AlertInstance<InstanceState, InstanceContext>>;
+}
+
+function trackAlertDurations<
+  InstanceState extends AlertInstanceState,
+  InstanceContext extends AlertInstanceContext
+>(params: TrackAlertDurationsParams<InstanceState, InstanceContext>) {
+  const currentTime = new Date().toISOString();
+  const { currentAlerts, originalAlerts, recoveredAlerts } = params;
+  const originalAlertIds = Object.keys(originalAlerts);
+  const currentAlertIds = Object.keys(currentAlerts);
+  const recoveredAlertIds = Object.keys(recoveredAlerts);
+  const newAlertIds = without(currentAlertIds, ...originalAlertIds);
+
+  // Inject start time into instance state of new instances
+  for (const id of newAlertIds) {
+    const state = currentAlerts[id].getState();
+    currentAlerts[id].replaceState({ ...state, start: currentTime, uuid: uuidv4() });
+  }
+
+  // Calculate duration to date for active instances
+  for (const id of currentAlertIds) {
+    const state = originalAlertIds.includes(id)
+      ? originalAlerts[id].getState()
+      : currentAlerts[id].getState();
+    const duration = state.start
+      ? (new Date(currentTime).valueOf() - new Date(state.start as string).valueOf()) * 1000 * 1000 // nanoseconds
+      : undefined;
+    currentAlerts[id].replaceState({
+      ...state,
+      ...(state.start ? { start: state.start } : {}),
+      ...(state.uuid ? { uuid: state.uuid } : {}),
+      ...(duration !== undefined ? { duration } : {}),
```

Comment:
This happens because we're just writing out a separate `new-instance` event for the first `active-instance` of an alert, so it contains the same information. I think it's ok for now.